matrix-meetings-bot/src/shared/calendarUtils/extractCalendarChange.ts

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright 2023 Nordeck IT + Consulting GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { differenceWith, isEqual } from 'lodash';
+import { CalendarEntryDto, DateTimeEntryDto } from '../../dto/CalendarEntryDto';
+import { formatICalDate, parseICalDate } from '../dateTimeUtils';
+import { isRRuleEntry, isRRuleOverrideEntry } from './helpers';
+
+export type OccurrenceChange =
+  | AddChange
+  | UpdateChange
+  | DeleteChange
+  | ExdateChange;
+
+type OverrideChangeBase = {
+  changeType: string;
+
+  /**
+   * Override entry.
+   */
+  value: CalendarEntryDto & {
+    recurrenceId: DateTimeEntryDto;
+  };
+};
+
+type AddChange = OverrideChangeBase & {
+  changeType: 'add';
+
+  /**
+   * Overriden occurrence end date and time.
+   */
+  dtend: DateTimeEntryDto;
```

Comment:
This design feels a bit odd (-> I didn't understand at first why this is needed and that the recurrence rule would implicitly be the corresponding start date). Would it be an option to directly calculate the start&end values for old&new that are needed for the later message at this place? It could still stay in the `DateTimeEntryDto` form if needed or as a ISO string or `DateTime` object.

Reply:
I changed this to be more explicit, please have a look.